util/src/rlp/rlpcompression.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright 2015, 2016 Ethcore (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use rlp::{UntrustedRlp, View, Compressible, encode, ElasticArray1024, Stream, RlpStream};
+use rlp::commonrlps::{BLOCKS_RLP_SWAPPER, SNAPSHOT_RLP_SWAPPER};
+use std::collections::HashMap;
+
+/// Stores RLPs used for compression
+pub struct InvalidRlpSwapper<'a> {
+	invalid_to_valid: HashMap<&'a [u8], &'a [u8]>,
+	valid_to_invalid: HashMap<&'a [u8], &'a [u8]>,
+}
+
+impl<'a> InvalidRlpSwapper<'a> {
+	/// Construct a swapper from a list of common RLPs
+	pub fn new(rlps_to_swap: &[&'a [u8]], invalid_rlps: &[&'a [u8]]) -> Self {
+		if rlps_to_swap.len() > 0x7e {
+			panic!("Invalid usage, only 127 RLPs can be swappable.");
+		}
+		let mut invalid_to_valid = HashMap::new();
+		let mut valid_to_invalid = HashMap::new();
+		for (&rlp, &invalid) in rlps_to_swap.iter().zip(invalid_rlps.iter()) {
+			invalid_to_valid.insert(invalid, rlp);
+			valid_to_invalid.insert(rlp, invalid);
+		}
+		InvalidRlpSwapper {
+			invalid_to_valid: invalid_to_valid,
+			valid_to_invalid: valid_to_invalid
+		}
+	}
+	/// Get a valid RLP corresponding to an invalid one
+	fn get_valid(&self, invalid_rlp: &[u8]) -> Option<&[u8]> {
+		self.invalid_to_valid.get(invalid_rlp).map(|r| r.clone())
+	}
+	/// Get an invalid RLP corresponding to a valid one
+	fn get_invalid(&self, valid_rlp: &[u8]) -> Option<&[u8]> {
+		self.valid_to_invalid.get(valid_rlp).map(|r| r.clone())
+	}
+}
+
+#[test]
+fn invalid_rlp_swapper() {
+	let to_swap: &[&[u8]] = &[&[0x83, b'c', b'a', b't'], &[0x83, b'd', b'o', b'g']];
+	let invalid_rlp: &[&[u8]] = &[&[0x81, 0x00], &[0x81, 0x01]];
+	let swapper = InvalidRlpSwapper::new(to_swap, invalid_rlp);
+	assert_eq!(Some(invalid_rlp[0]), swapper.get_invalid(&[0x83, b'c', b'a', b't']));
+	assert_eq!(None, swapper.get_invalid(&[0x83, b'b', b'a', b't']));
+	assert_eq!(Some(to_swap[1]), swapper.get_valid(invalid_rlp[1]));
+}
+
+/// Type of RLP indicating its origin database.
+pub enum RlpType {
```

Comment:
I don't like this, should it be a field of `UntrustedRlp` which indicates origin.